### PR TITLE
Enable `uclamp` configs to save oodles of battery life and update some kernel configs

### PR DESCRIPTION
**Pull request by mkurz**
**First commit:**
Enables [uclamp](https://docs.kernel.org/scheduler/sched-util-clamp.html), which is necessary that `speakersafetd` will be "pushed" to the e (efficiency) cpu cores, which safes lots of battery (you feel the difference). I got that hint from [this chat log](https://oftc.irclog.whitequark.org/asahi/2024-02-12#1707762141-1707762277;):
```
[18:22](https://oftc.irclog.whitequark.org/asahi/2024-02-12#32916743) <milky> I am getting this error when the speakersafetydeamon tries to run
[18:22](https://oftc.irclog.whitequark.org/asahi/2024-02-12#32916744) <milky> Could not lock elem Left Woofer 1 VSENSE Switch. alsa-lib error: Error("snd_ctl_elem_lock", ENOENT)
[18:22](https://oftc.irclog.whitequark.org/asahi/2024-02-12#32916747) <milky> also "Failed to set uclamp"
[18:23](https://oftc.irclog.whitequark.org/asahi/2024-02-12#32916748) <j`ey> milky: youre probably missing the patches that actually enables the speakers see branch: speakers/enablement-READ-COMMIT-MESSAGE
[18:23](https://oftc.irclog.whitequark.org/asahi/2024-02-12#32916749) <j`ey> and for the latter, you need CONFIG_UCLAMP_TASKS
[18:24](https://oftc.irclog.whitequark.org/asahi/2024-02-12#32916752) <j`ey> and UCLAMP_TASK_GROUP
```

Then I remembered I did read about that somewhere... which was the [Asahi Blog entry here](https://asahilinux.org/2024/01/fedora-asahi-new/#energy-aware-scheduling) :wink: (also marcan [wrote about it](https://social.treehouse.systems/@marcan/111363323102625920))
> ... While working on speaker support, we found that Pipewire and Wireplumber were constantly being mis-scheduled onto P-cores. ... To fix this, we gave Pipewire and Wireplumber the ability to use utilisation clamping, a scheduler feature ... We cap Pipewire and Wireplumber to an extremely low maximum performance so the scheduler restricts them to efficiency cores at their lowest operating point. .... we get to save oodles of battery life for our users! ... **it wasn’t even enabled in the standard Fedora kernel until we asked for it to be a couple of weeks ago (`CONFIG_UCLAMP_TASK`)**...

E.g., currently, when `speakersafetd` starts up you get following error:
```
$ journalctl -u speakersafetyd.service --grep uclamp
Mai 17 14:23:40 mkurz-macbook-pro speakersafetyd[664]: WARN  [speakersafetyd::uclamp] Failed to set uclamp
```
With uclamp enabled `speakersafetd` works correctly:
```
Mai 17 14:27:01 mkurz-macbook-pro speakersafetyd[672]: INFO  [speakersafetyd::uclamp] Set task uclamp to 0:64
```

Precious battery life saved! :rocket: 

See:
- https://github.com/AsahiLinux/asahi-audio/blob/v2.2/conf/pipewire.conf#L15-L16
- https://github.com/AsahiLinux/asahi-audio/blob/v2.2/conf/wireplumber.conf#L9-L10
- https://github.com/AsahiLinux/speakersafetyd/commit/c3307388aa72e43af3bf6d6b2498f02a8016e2b1
- https://github.com/AsahiLinux/speakersafetyd/blob/0.1.9/src/main.rs#L148-L151 :point_down: 
  - :point_right:  https://github.com/AsahiLinux/speakersafetyd/blob/0.1.9/src/uclamp.rs#L18-L50

**Second commit:**
Updates some configs according to wiki and Fedora config:
- https://github.com/AsahiLinux/docs/wiki/Kernel-config-notes-for-distros
- https://copr-dist-git.fedorainfracloud.org/cgit/@asahi/kernel/kernel.git/tree/kernel-aarch64-16k-fedora.config?h=f40
- BTW, also referencing the asahi ubuntu kernelf configs:
  - https://github.com/UbuntuAsahi/linux/blob/master/debian.apple-arm/config/annotations (the asahi specific one)
  - https://github.com/UbuntuAsahi/linux/blob/master/debian.master/config/annotations (the common one)

It seems the wiki recommends to include most configs instead of building them as module. @joske This is up to you, please review if we should keep the configs modules or not.
I will comment separately on configs which have not been enabled yet.

**Third commit:**
I saw that `CONFIG_RUST_PHYLIB_ABSTRACTIONS` is enabled in both the Fedora and the Ubuntu kernel, so I thought because this is rust related maybe we should enable it too?
- https://copr-dist-git.fedorainfracloud.org/cgit/@asahi/kernel/kernel.git/tree/kernel-aarch64-16k-fedora.config?h=f40#n6838
- https://github.com/UbuntuAsahi/linux/blob/master/debian.apple-arm/config/annotations#L99

What do you think?

Of course I tested this config on my MacBook Pro (14-inch, M1 Pro, 2021) (J314sAP) and it works.

**BTW:**
While working on this I found out the asahi wiki page is not up to date for the 6.8 kernel config:
- https://github.com/zzywysm/asahi-kernel-configs/issues/1 (this GitHub users seems to be the one maintaining that wiki page)

### CODE REVIEW

File: linux-asahi/config [hunk 1]
```diff
@@ -185,6 +186,7 @@ CONFIG_FAIR_GROUP_SCHED=y
 CONFIG_CFS_BANDWIDTH=y
 # CONFIG_RT_GROUP_SCHED is not set
 CONFIG_SCHED_MM_CID=y
+CONFIG_UCLAMP_TASK_GROUP=y
```
Comment (mkurz) [inline linux-asahi/config:189]: The above three enable uclamp, just like Fedora and ubuntu does:
- https://copr-dist-git.fedorainfracloud.org/cgit/@asahi/kernel/kernel.git/tree/kernel-aarch64-16k-fedora.config?h=f40#n8953
- https://github.com/UbuntuAsahi/linux/blob/master/debian.master/config/annotations#L14138-L14140

Also recommended in the wiki:
- https://github.com/AsahiLinux/docs/wiki/Kernel-config-notes-for-distros

File: linux-asahi/config [hunk 2]
```diff
@@ -271,9 +273,9 @@ CONFIG_PERF_EVENTS=y
 CONFIG_SYSTEM_DATA_VERIFICATION=y
 CONFIG_PROFILING=y
 CONFIG_RUST=y
-CONFIG_RUSTC_VERSION_TEXT="rustc 1.76.0 (07dca489a 2024-02-04)"
+CONFIG_RUSTC_VERSION_TEXT="rustc 1.78.0 (9b00956e5 2024-04-29) (Arch Linux rust 1:1.78.0-1)"
 CONFIG_RUSTC_LLVM_VERSION_TEXT="17.0.6"
-CONFIG_BINDGEN_VERSION_TEXT="bindgen 0.65.1"
+CONFIG_BINDGEN_VERSION_TEXT="bindgen 0.69.4"
```
Comment (mkurz) [inline linux-asahi/config:278]: These two text values changed automatically by `make menuconfig`.
@joske Should we keep it?

File: linux-asahi/config [hunk 3]
```diff
@@ -750,8 +752,7 @@ CONFIG_ARCH_HAS_GCOV_PROFILE_ALL=y
 # end of GCOV-based kernel profiling
 
 CONFIG_HAVE_GCC_PLUGINS=y
-CONFIG_GCC_PLUGINS=y
-# CONFIG_GCC_PLUGIN_LATENT_ENTROPY is not set
+# CONFIG_GCC_PLUGINS is not set
```
Comment (mkurz) [inline linux-asahi/config:755]: Wiki says not to set `CONFIG_GCC_PLUGINS`:
- https://github.com/AsahiLinux/docs/wiki/Kernel-config-notes-for-distros

Also not set in Fedora and Ubuntu:
- https://copr-dist-git.fedorainfracloud.org/cgit/@asahi/kernel/kernel.git/tree/kernel-aarch64-16k-fedora.config?h=f40#n2680

File: linux-asahi/config [hunk 4]
```diff
@@ -2692,7 +2693,7 @@ CONFIG_SWPHY=y
 CONFIG_LED_TRIGGER_PHY=y
 CONFIG_PHYLIB_LEDS=y
 CONFIG_FIXED_PHY=y
-# CONFIG_RUST_PHYLIB_ABSTRACTIONS is not set
+CONFIG_RUST_PHYLIB_ABSTRACTIONS=y
```
Comment (mkurz) [inline linux-asahi/config:2696]: Already mentiond in the initial commit.

File: linux-asahi/config [hunk 5]
```diff
@@ -3149,7 +3151,80 @@ CONFIG_TABLET_USB_HANWANG=m
 CONFIG_TABLET_USB_KBTAB=m
 CONFIG_TABLET_USB_PEGASUS=m
 # CONFIG_TABLET_SERIAL_WACOM4 is not set
-# CONFIG_INPUT_TOUCHSCREEN is not set
+CONFIG_INPUT_TOUCHSCREEN=y
```
Comment (mkurz) [inline linux-asahi/config:3154]: Needed for `CONFIG_TOUCHSCREEN_APPLE_Z2`

File: linux-asahi/config [hunk 6]
```diff
@@ -3149,7 +3151,80 @@ CONFIG_TABLET_USB_HANWANG=m
...
+# CONFIG_TOUCHSCREEN_ADS7846 is not set
+# CONFIG_TOUCHSCREEN_AD7877 is not set
+# CONFIG_TOUCHSCREEN_AD7879 is not set
+CONFIG_TOUCHSCREEN_APPLE_Z2=m
```
Comment (mkurz) [inline linux-asahi/config:3158]: Recommend in the wiki and build as module in Fedora and Asahi as well:
- https://github.com/AsahiLinux/docs/wiki/Kernel-config-notes-for-distros
- https://copr-dist-git.fedorainfracloud.org/cgit/@asahi/kernel/kernel.git/tree/kernel-aarch64-16k-fedora.config?h=f40#n8765
- https://github.com/UbuntuAsahi/linux/blob/master/debian.apple-arm/config/annotations#L112

File: linux-asahi/config [hunk 7]
```diff
@@ -4920,6 +4995,7 @@ CONFIG_VIDEO_CMDLINE=y
 CONFIG_VIDEO_NOMODESET=y
 # CONFIG_AUXDISPLAY is not set
 CONFIG_DRM=y
+CONFIG_DRM_MIPI_DSI=y
```
Comment (mkurz) [inline linux-asahi/config:4998]: I think this was enabled as dependency automatically... Also enabled by ubuntu anyway:
- https://github.com/UbuntuAsahi/linux/blob/master/debian.master/config/annotations#L4190

File: linux-asahi/config [hunk 8]
```diff
@@ -5078,7 +5211,7 @@ CONFIG_DRM_SIMPLEDRM=y
 # CONFIG_DRM_LIMA is not set
 # CONFIG_DRM_PANFROST is not set
 # CONFIG_DRM_TIDSS is not set
-# CONFIG_DRM_ADP is not set
+CONFIG_DRM_ADP=m
```
Comment (mkurz) [inline linux-asahi/config:5214]: Recommended by wiki and enabled in Fedora and Ubuntu:
- https://github.com/AsahiLinux/docs/wiki/Kernel-config-notes-for-distros
- https://copr-dist-git.fedorainfracloud.org/cgit/@asahi/kernel/kernel.git/tree/kernel-aarch64-16k-fedora.config?h=f40#n1913
- https://github.com/UbuntuAsahi/linux/blob/master/debian.apple-arm/config/annotations#L61

File: linux-asahi/config [hunk 9]
```diff
@@ -6826,7 +6959,7 @@ CONFIG_MULTIPLEXER=m
 #
 # CONFIG_MUX_ADG792A is not set
 # CONFIG_MUX_ADGS1408 is not set
-# CONFIG_MUX_APPLE_DPXBAR is not set
+CONFIG_MUX_APPLE_DPXBAR=m
```
Comment (mkurz) [inline linux-asahi/config:6962]: Again recommended in wiki and enabled in Fedora and Ubuntu

File: linux-asahi/config (hunk 7 above)
Comment (joske) [reply linux-asahi/config:4998]: this looks unnecessary

File: linux-asahi/config (hunk 2 above)
Comment (joske) [reply linux-asahi/config:278]: this does not matter much, it's only informational

File: linux-asahi/config (hunk 4 above)
Comment (joske) [reply linux-asahi/config:2696]: it's not mentioned in the wiki, but if fedora and ubuntu do this, let's add it too

**Comment by joske** [comment]
@mkurz Thanks for taking the time to go over this! 

Can you sign off your commits please?

File: linux-asahi/config (hunk 7 above)
Comment (mkurz) [reply linux-asahi/config:4998]: I can not remove it, it seems it's enabled automatically as dependency. Even if I remove it from the config file by hand and then re-run `make menuconfig` it gets added again. Also look here, it's selected by others (and I can not switch to it):
image

**Comment by zzywysm** [comment]
> It seems the wiki recommends to include most configs instead of building them as module. @joske This is up to you, please review if we should keep the configs modules or not. I will comment separately on configs which have not been enabled yet.

If you are building a general arm64 16k kernel for Arch Linux ARM intended to be bootable on any arm64 16k hardware (much like how Fedora does it), you will want to build as many Asahi configuration options as modules as you can.

If you are building a kernel that's only ever intended to run on Apple Silicon Macs, I think it makes a lot more sense to build more of the Asahi code into the kernel image directly.  My kernel config recommendations on the Asahi wiki reflect this opinion.

**Pull request merged**